### PR TITLE
[Feature] Request Table API Hookup

### PR DESCRIPTION
**Pull request by angeladietz**
## Notion ticket link

[Requests API hookup](https://www.notion.so/uwblueprintexecs/Requests-API-hookup-ec81141ad07c48dbaf8bf1287f687420)

## Implementation description
* Hooks up API for request table to display appropriate applications
* supports filtering, pagination, sorting, searching
* added `permitType` field to `Applications` DB table
* created `PermitType` enum
* Added more values to `ApplicationStatus` enum

## Notes
* We might need to fix a small UX bug with the sorting -- the arrow disappears once the data refreshes and will only reappear if we click somewhere on the header
* This PR will also probably conflict with #42 

## Checklist
- [x] My PR name is descriptive, is in imperative tense and starts with one of the following: `[Feature]`,`[Improvement]` or `[Fix]`,
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the RCD team on GitHub, or specific people who are associated with this ticket

### CODE REVIEW

File: lib/applications/resolvers.ts
Comment (EmilioMena) [inline lib/applications/resolvers.ts:66]: So you don't have to repeat this in both the count query and actual query you can take out the `where` in a variable and pass it to both queries. it will also be easier to make a change in one place in the future if needed

File: lib/applications/resolvers.ts
Comment (EmilioMena) [inline lib/applications/resolvers.ts:102]: Is there a reason we're explicitly selecting only these columns? I don't think we should constrain which fields are being returned. Graphql already handles selecting only the data that we need in the query.

File: lib/applications/schema.graphql
Comment (EmilioMena) [inline lib/applications/schema.graphql:163]: I don't think we need a custom return type when filtering applications, `Application` should work just fine, no?
applicationProcessing should also probably be added to the Application type as well

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:63]: hm shouldn't there always be an ID for an application?

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:163]: is middle name also needed here?

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:164]: ah it looks like we're using the database id and not the rcdUserId here, probably need to change that

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:128]: I had this as 1 for testing too lol just change it back to 20 before merging

**Review by EmilioMena — COMMENTED** [review]
Awesome work so far y'all! Left down some comments

File: lib/applications/resolvers.ts
Comment (angeladietz) [reply lib/applications/resolvers.ts:102]: Good point. we had to specifically select the applicationProcessing field but I realized we can use `include` instead so I switched to that

File: lib/applications/schema.graphql
Comment (angeladietz) [reply lib/applications/schema.graphql:163]: Yeah you're right, good catch

File: pages/admin/index.tsx
Comment (angeladietz) [reply pages/admin/index.tsx:63]: The ID is the applicant ID, not the application ID, so if someone is applying for the first time, they won't have an applicant ID yet.

File: pages/admin/index.tsx
Comment (angeladietz) [reply pages/admin/index.tsx:163]: We're not showing the middle name on the table (which follows the design) so it's left out.

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:360]: Oustan and I added a [useDebounce](https://usehooks.com/useDebounce/) hook in my [latest commit](https://github.com/uwblueprint/richmond-centre-for-disability/pull/42/commits/94abb4044c7b17244d7adf3e8532ce6983f8a9af) so that we only need to use onChange here. The debouncer will only set the searchFilter state if it's value hasn't been changed in the last 500ms. That way we can avoid making users press enter to search without firing a db query every time a key is pressed. It's better ux and looks cooler imo

Can you copy those changes in here?

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:377]: also fixed a bug in the pagination component in my latest commit. It now sets the page number in the props instead of keeping the logic inside the component which allows the pageNumber to be set to 0 after a filter change

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:332]: can you group all the setPageNumber(0) in a useEffect hook? That it doesn't have to be repeated for every filter

File: pages/admin/index.tsx
Comment (EmilioMena) [inline pages/admin/index.tsx:195]: Do we need sortOrder here?

File: tools/pages/admin/index.ts
Comment (EmilioMena) [inline tools/pages/admin/index.ts:30]: remove comment

**Review by EmilioMena — APPROVED** [review]
Awesome job! Left a couple small questions/comments but other than that good to go 🚢

File: pages/admin/index.tsx
Comment (angeladietz) [reply pages/admin/index.tsx:195]: Yeah we do assuming that when the sorting order changes we want to go back to the first page (which I assume is the case because the data on the current page will most likely be different).

File: lib/applications/resolvers.ts
Comment (OustanDing) [inline lib/applications/resolvers.ts:28]: Nice documentation 🤩

File: lib/applications/resolvers.ts
Comment (OustanDing) [inline lib/applications/resolvers.ts:57]: Can use `SortOptions` from `tools/types.ts`

File: lib/applications/schema.graphql
Comment (OustanDing) [inline lib/applications/schema.graphql:66]:
```suggestion
  createdAt: Date!
```

File: lib/applications/schema.graphql
Comment (OustanDing) [inline lib/applications/schema.graphql:155]:
```suggestion
  id: Int!
  status: ApplicationStatus!
  appNumber: Int
  appHolepunched: Boolean!
  walletCardCreated: Boolean!
  invoiceNumber: Int
  documentUrls: [String!]
  appMailed: Boolean!
  applicationId: Int!
  createdAt: Date!
  updatedAt: Date!
  application: Application!
```

File: lib/graphql/schema.graphql
Comment (OustanDing) [inline lib/graphql/schema.graphql:109]:
```suggestion
enum PermitType {
```

File: lib/graphql/schema.graphql
Comment (OustanDing) [inline lib/graphql/schema.graphql:89]: This enum is used for the application processing steps, so it should only contain `PENDING`, `APPROVED`, `REJECTED` and `COMPLETED`.

`EXPIRING`, `EXPIRED` and `ACTIVE` are computed dynamically. `INPROGRESS` is equal to `APPROVED` (if an application is approved but not yet completed, it is currently being processed).

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:64]: Similar to in @EmilioMena 's PR, I think we can just put all of these render functions in the columns configuration to keep it all in one place. Also, I believe it can infer the types if you do it like that.

File: tools/pages/admin/index.ts
Comment (OustanDing) [inline tools/pages/admin/index.ts:42]:
```suggestion
    result: ReadonlyArray<
      Pick<
        Application,
        | 'firstName'
        | 'lastName'
        | 'id'
        | 'createdAt'
        | 'permitType'
        | 'isRenewal'
        | 'applicantId'
      > & {
        applicationProcessing: Pick<ApplicationProcessing, 'status'>
      }
    >;
```
We also want to pick from the `applicationProcessing` object

File: pages/admin/index.tsx
Comment (OustanDing) [reply pages/admin/index.tsx:195]: Agreed

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:151]: We don't need a state variable for this -  the page and table will rerender when any of the filters change, so the table data will be updated automatically as well.

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:153]: Also don't need a state variable for `recordsCount`

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:156]: Following the two comments above, we can grab the query result like this:
```suggestion
  const { data } = useQuery<GetApplicationsResponse, GetApplicationsRequest>(GET_APPLICATIONS_QUERY, {
```
and we can destructure the `results` and `totalCount` from `data`

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:180]: We want to move this table data formatting logic to the `Cell` parameter in the column configuration just to keep everything in one place.

**Review by OustanDing — CHANGES_REQUESTED** [review]
Great work! Left some comments, mostly regarding consistency. Functionality and logic look sound. After that, we should be good to merge :)

File: lib/applications/resolvers.ts
Comment (jennifertsai) [reply lib/applications/resolvers.ts:57]: If we use the `SortOptions` instead, we can't pass it into Prisma directly for the `orderBy` field

Comment (angeladietz) [reply lib/applications/resolvers.ts:57]: ^ yeah, `order` is already a `SortOptions` so here we're parsing it into a format that we can pass into the prisma query.

File: pages/admin/index.tsx
Comment (angeladietz) [reply pages/admin/index.tsx:174]: We actually can't do this because the data value is never used. The `data` that's used inside onCompleted doesn't correspond to the same data variable so it produces errors

File: lib/applications/resolvers.ts
Comment (OustanDing) [reply lib/applications/resolvers.ts:57]: Gotcha, thanks!

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:45]: Tiny nit. Name is misleading because these aren't component props. Easiest fix would be to just remove this type.

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:57]:
```suggestion
          <Text>{`${value.firstName} ${value.lastName}`}</Text>
```
JS string interpolation

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:120]:
```suggestion
    rcdUserId?: number;
```

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:125]:
```suggestion
  status: ApplicationStatus?;
```

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:176]:
```suggestion
          status: record.applicationProcessing?.status,
```
The `applicationProcessing?` should cover the `undefined` case

File: pages/admin/index.tsx
Comment (OustanDing) [reply pages/admin/index.tsx:147]: Note for future: Discussed with @angeladietz and we will keep using the state variable as it's necessary for table rerendering. My bad :/

**Review by OustanDing — COMMENTED** [review]
Few small nits. Sorry I missed them in the previous review

File: pages/admin/index.tsx
Comment (OustanDing) [inline pages/admin/index.tsx:40]: Just wondering why this was changed? I think it should be working fine

**Review by OustanDing — APPROVED** [review]
Fixes look good! I just left one question in the review. Will approve since everything else looks great. Thanks so much for your great work @angeladietz @jennifertsai !

File: pages/admin/index.tsx
Comment (angeladietz) [reply pages/admin/index.tsx:40]: Without using the `PermitTypeProps`, it interprets the result as an `any` and doesn't allow for reading from `permitTypeString` since it might not be a `PermitType`.
[Exhibit A](https://railway.app/project/afc837ea-aa06-4d86-8732-287740d12581/deployments?id=3b3e73a7-e9cc-4828-8ee5-e2d600b5f5c0)
[Exhibit B](https://railway.app/project/afc837ea-aa06-4d86-8732-287740d12581/deployments?id=946771c2-4188-4a4c-97bf-a6cd156f95e0)
[Exhibit C](https://railway.app/project/afc837ea-aa06-4d86-8732-287740d12581/deployments?id=1f4921c8-40d4-4a76-a3dd-988014a4469c)

If you have a better fix then lmk and we can address it separately

Comment (OustanDing) [reply pages/admin/index.tsx:40]: Got it. Will add to backlog.

**Pull request merged**
